### PR TITLE
ddl: recover to the correct partition from checkpoint

### LINKED ISSUE #43997: Doing add index causes inconsistent data when DDL owner change (like upgrade the cluster)
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. deploy a cluster with version v7.1.0
2. load some data and add index
3. upgrade the cluster 
4. admin check table
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
admin check table will pass
### 3. What did you see instead (Required)
```sql
ERROR 8223 (HY000): data inconsistency in table: t, index: idx1, handle: 3831699, index-values:"" != record-values:"handle: 3831699, values: [KindMysqlTime 2023-05-18 16:36:41 KindMysqlTime 2023-05-18 16:36:41 KindString b0d736f6-bc4e-4dba-958e-fab37cbc144a-1684399001651058794 KindString 3051fa5f-fb6a-4de6-a0df-39c88df91938305 KindString 2 KindString bd555c0d-ea74-4156-9633-51c2434537bfbd555c0d-ea74-4156-9633-51c24345 KindString f532e9e5-69a4-4b KindString 39147ffa-4a91-4c48-8f5a-b6ac6da3befd39147f KindStri
```
### 4. What is your TiDB version? (Required)
before 
```sql
Release Version: v7.1.0
Edition: Community
Git Commit Hash: 705553fa06b92fcaeb34957d1db521c10da9358c
Git Branch: heads/refs/tags/v7.1.0
UTC Build Time: 2023-05-18 02:44:40
GoVersion: go1.20.3
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
```
after 
```sql
MySQL [rtdb]> select tidb_version();
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                           |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v7.2.0-alpha
Edition: Community
Git Commit Hash: 2c2730069e3ace7a6763f010445354b4ac9db2f2
Git Branch: heads/refs/tags/v7.2.0-alpha
UTC Build Time: 2023-05-18 03:45:55
GoVersion: go1.20.3
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
```
<!-- Paste the output of SELECT tidb_version() -->

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #43997

Problem Summary:

The basic idea of checkpoint is to recover the progress:

```
| add index for partition 1
|  [ local checkpoint ]
| add index for partition 2
|  [ local checkpoint ]
|  [ global checkpoint ]
| ...
| add index for partition k
|  [ local checkpoint ]
| add index for partition k+1
v  (TiDB 1) crash
  | (TiDB 2) get DDL owner
  | add index for partition 2
  | ...
```

Note that we can only begin with partition 2 because the local checkpoint is lost when TiDB 1 crashes.

In order to represent which partition we should begin with, reorg meta is used. The reorg meta contains a tuple: (partition ID or physical table ID, start key, end key). Every time TiDB restarts in the middle state of adding an index, it tries to reset the reorg meta to the state exactly before the last global checkpoint.

Previously, we store the reorg meta in the checkpoint manager. However, we did not distinguish the "local" reorg meta and the "global" reorg meta. When a partition is complete, the reorg meta is updated immediately, leading to a new TiDB reset to the wrong partition. Finally, the index data from some of the partitions is lost.

### What is changed and how it works?

- Distinguish the "local" reorg meta from the "global" one.
- When the `mysql.ddl_reorg_meta` is initialized, we also initialize the checkpoint.
- Move the creation of checkpoint manager to a proper place(which needs the info from `mysql.ddl_reorg_meta`).
- After the checkpoint manager is created, we try recover the global checkpoint and overwrite the reorg info.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  ```
  1. Create two TiDB.
  2. Prepare a table with a lot of partitions.
  3. Add index.
  4. Kill the TiDB owner. 
  5. Check if the other TiDB can reset the reorg meta to a correct partition ID. 
  ```
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: ddl/job_table.go [hunk 1]
```diff
@@ -632,9 +635,19 @@ func updateDDLReorgHandle(se *sess.Session, jobID int64, startKey kv.Key, endKey
...
+			EndKey:     endKey,
+			Version:    ingest.JobCheckpointVersionCurrent,
+		}})
+	if err != nil {
+		return errors.Trace(err)
+	}
 	del := fmt.Sprintf("delete from mysql.tidb_ddl_reorg where job_id = %d", jobID)
-	ins := fmt.Sprintf("insert into mysql.tidb_ddl_reorg(job_id, ele_id, ele_type, start_key, end_key, physical_id) values (%d, %d, %s, %s, %s, %d)",
-		jobID, element.ID, util.WrapKey2String(element.TypeKey), util.WrapKey2String(startKey), util.WrapKey2String(endKey), physicalTableID)
+	ins := fmt.Sprintf("insert into mysql.tidb_ddl_reorg(job_id, ele_id, ele_type, start_key, end_key, physical_id, reorg_meta) values (%d, %d, %s, %s, %s, %d, %s)",
```
Comment (zimulala) [inline ddl/job_table.go:649]: Do we need to add `reorg_meta` to `UpdateReorgMeta`?

Comment (tangenta) [reply ddl/job_table.go:649]: No, the field `reorg_meta` is updated by the checkpoint manager:

https://github.com/pingcap/tidb/blob/master/ddl/ingest/checkpoint.go#L319

File: ddl/reorg.go [hunk 2]
```diff
@@ -322,7 +317,30 @@ func (w *worker) runReorgJob(rh *reorgHandler, reorgInfo *reorgInfo, tblInfo *mo
 	return nil
 }
 
-func overwriteReorgInfoFromCheckpoint(sess *sess.Session, job *model.Job, reorgInfo *reorgInfo) error {
+func overwriteReorgInfoFromGlobalCheckpoint(w *worker, sess *sess.Session, job *model.Job, reorgInfo *reorgInfo) error {
```
Comment (wjhuang2016) [inline ddl/reorg.go:320]: It should not handle distributed execution.

File: ddl/ingest/checkpoint.go [hunk 3]
```diff
@@ -174,18 +178,23 @@ func (s *CheckpointManager) UpdateCurrent(taskID int, added int) error {
...
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	s.progressLocalSyncMinKey()
 	if imported && s.minKeySyncGlobal.Cmp(s.minKeySyncLocal) != 0 {
 		s.minKeySyncGlobal = s.minKeySyncLocal
 		s.globalCnt = s.localCnt
 		s.dirty = true
+
+		s.pidGlobal = s.pidLocal
```
Comment (zimulala) [inline ddl/ingest/checkpoint.go:190]: Why here we need to update global info? It seems we do it after `WriteLocal`, so why not update local info?

File: ddl/job_table.go (hunk 1 above)
Comment (zimulala) [reply ddl/job_table.go:649]: Got it.

File: ddl/ingest/checkpoint.go (hunk 3 above)
Comment (zimulala) [reply ddl/ingest/checkpoint.go:190]: Do import to tikv in https://github.com/pingcap/tidb/pull/44024/files#diff-451edd897e959340ba4a6565b82969ae58c35a4b9715674769565f3958c39a2dR177.

File: ddl/reorg.go (hunk 2 above)
Comment (tangenta) [reply ddl/reorg.go:320]: Updated.

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**